File: scala/ltx/src/test/scala/SecurityTest.scala
/**
 * SecurityTest.scala -- Epic 29, Stories 29.1 / 29.4 / 29.5
 */

object SecurityTest:
  var passed = 0
  var failed = 0

  def check(label: String, cond: Boolean): Unit =
    if cond then passed += 1
    else { failed += 1; println(s"FAIL: $label") }

  def checkEq(label: String, got: Any, exp: Any): Unit =
    if got == exp then passed += 1
    else { failed += 1; println(s"FAIL: $label  expected=$exp  got=$got") }

  def main(args: Array[String]): Unit =
    // ---- canonical_json ----
    checkEq("empty object", Security.canonicalJson(Map.empty[String,Any]), "{}")
    checkEq("sorted keys", Security.canonicalJson(Map("z" -> 1, "a" -> 2)), "{\"a\":2,\"z\":1}")
    checkEq("array", Security.canonicalJson(Seq(1, 2, 3)), "[1,2,3]")
    checkEq("number", Security.canonicalJson(42), "42")
    checkEq("bool", Security.canonicalJson(true), "true")
    checkEq("string", Security.canonicalJson("hi"), "\"hi\"")
    checkEq("null", Security.canonicalJson(null), "null")
    val nested = Security.canonicalJson(Map("b" -> Map("y" -> 9, "x" -> 1), "a" -> 3))
    checkEq("nested sorted", nested, "{\"a\":3,\"b\":{\"x\":1,\"y\":9}}")

    // ---- generate_nik ----
    val nik1 = Security.generateNik()
    val nik2 = Security.generateNik()
    checkEq("key_type", nik1.keyType, "ltx-nik-v1")
    check("node_id non-empty", nik1.nodeId.nonEmpty)
    check("kid non-empty", nik1.kid.nonEmpty)
    checkEq("node_id 22 chars", nik1.nodeId.length, 22)
    check("pub_key non-empty", nik1.publicKeyB64.nonEmpty)
    check("priv_key non-empty", nik1.privateKeyB64.nonEmpty)
    check("issued_at set", nik1.issuedAt.nonEmpty)
    check("expires_at set", nik1.expiresAt.nonEmpty)
    val nikLbl = Security.generateNik(validDays = 30, nodeLabel = "TestNode")
    checkEq("node_label", nikLbl.nodeLabel, "TestNode")
    check("expires after issued", nikLbl.expiresAt > nikLbl.issuedAt)
    check("unique node_ids", nik1.nodeId != nik2.nodeId)

    // ---- is_nik_expired ----
    check("fresh nik not expired", !Security.isNikExpired(nik1))
    val oldNik = nik1.copy(expiresAt = "2000-01-01T00:00:00Z")
    check("old nik expired", Security.isNikExpired(oldNik))

    // ---- sign_plan / verify_plan ----
    val plan = Map("planId" -> "p1", "startAt" -> "2026-05-01T00:00:00Z", "quantum" -> 60)
    val sp   = Security.signPlan(plan, nik1)
    check("coseSign1 protected", sp.coseSign1.protectedHdr.nonEmpty)
    check("coseSign1 signature", sp.coseSign1.signature.nonEmpty)
    checkEq("kid in unprotected", sp.coseSign1.kid, nik1.kid)

    val cache = Map(nik1.kid -> nik1)
    val (ok1, r1) = Security.verifyPlan(sp, cache)
    check("verify ok", ok1)
    checkEq("verify reason", r1, "ok")

    val (ok2, r2) = Security.verifyPlan(sp, Map.empty)
    check("verify fails empty cache", !ok2)
    checkEq("reason key_not_in_cache", r2, "key_not_in_cache")

    val expiredCache = Map(nik1.kid -> oldNik)
    val (ok3, r3) = Security.verifyPlan(sp, expiredCache)
    check("verify fails expired key", !ok3)
    checkEq("reason key_expired", r3, "key_expired")

    val tampered = sp.copy(plan = Map("planId" -> "TAMPERED"))
    val (ok4, r4) = Security.verifyPlan(tampered, cache)
    check("verify fails tampered", !ok4)
    checkEq("reason payload_mismatch", r4, "payload_mismatch")

    // ---- SequenceTracker ----
    val st = new Security.SequenceTracker("plan-x")
    checkEq("plan_id stored", st.planId, "plan-x")

    val (r1a, m1a) = st.addSeq("alice", 1)
    check("first seq accepted", r1a)
    checkEq("first seq msg", m1a, "ok")

    val (r2a, m2a) = st.addSeq("alice", 2)
    check("seq 2 accepted", r2a)
    checkEq("seq 2 msg", m2a, "ok")

    val (r3a, m3a) = st.addSeq("alice", 2)
    check("replay rejected", !r3a)
    checkEq("replay msg", m3a, "replay")

    val (r4a, m4a) = st.addSeq("alice", 10)
    check("gap accepted", r4a)
    checkEq("gap msg", m4a, "gap")

    val (c1, _) = st.checkSeq("alice", 10)
    check("check_seq replay", !c1)
    val (c2, _) = st.checkSeq("alice", 11)
    check("check_seq next", c2)
    val (c3, mc) = st.checkSeq("alice", 20)
    check("check_seq gap", c3)
    checkEq("check_seq gap msg", mc, "gap")

    val (rb, mb) = st.addSeq("bob", 5)
    check("bob first seq", rb)
    checkEq("bob first msg", mb, "ok")

    println(s"\n$passed passed  $failed failed")
    if failed > 0 then System.exit(1)

File: ocaml/ltx/test/security_test.ml
(* security_test.ml -- Epic 29, Stories 29.1 / 29.4 / 29.5 *)

let passed = ref 0
let failed = ref 0

let check msg cond =
  if cond then incr passed
  else begin
    incr failed;
    Printf.printf "FAIL: %s\n" msg
  end

let check_eq msg a b =
  if a = b then incr passed
  else begin
    incr failed;
    Printf.printf "FAIL: %s  expected=%s  got=%s\n" msg b a
  end

let () =
  (* ---- canonical_json ---- *)
  let open Security in
  check_eq "empty object" (canonical_json (JObj [])) "{}";
  check_eq "sorted keys" (canonical_json (JObj [("z",JInt 1);("a",JInt 2)])) "{\"a\":2,\"z\":1}";
  check_eq "array" (canonical_json (JArr [JInt 1;JInt 2;JInt 3])) "[1,2,3]";
  check_eq "number" (canonical_json (JInt 42)) "42";
  check_eq "bool true" (canonical_json (JBool true)) "true";
  check_eq "string" (canonical_json (JStr "hi")) "\"hi\"";
  check_eq "null" (canonical_json JNull) "null";
  check_eq "nested" (canonical_json (JObj [("b",JObj [("y",JInt 9);("x",JInt 1)]);("a",JInt 3)]))
    "{\"a\":3,\"b\":{\"x\":1,\"y\":9}}";

  (* ---- generate_nik ---- *)
  let nik  = generate_nik () in
  let nik2 = generate_nik () in
  check "key_type" (nik.key_type = "ltx-nik-v1");
  check "node_id non-empty" (String.length nik.node_id > 0);
  check "kid non-empty" (String.length nik.kid > 0);
  check "node_id is 22 chars" (String.length nik.node_id = 22);
  check "pub_key non-empty" (String.length nik.public_key_b64 > 0);
  check "priv_key non-empty" (String.length nik.private_key_b64 > 0);
  check "issued_at set" (String.length nik.issued_at > 0);
  check "expires_at set" (String.length nik.expires_at > 0);
  let nik_lbl = generate_nik ~valid_days:30 ~node_label:"TestNode" () in
  check_eq "node_label" nik_lbl.node_label "TestNode";
  check "expires after issued" (nik_lbl.expires_at > nik_lbl.issued_at);
  check "unique node_ids" (nik.node_id <> nik2.node_id);

  (* ---- is_nik_expired ---- *)
  check "fresh nik not expired" (not (is_nik_expired nik));
  let old_nik = { nik with expires_at = "2000-01-01T00:00:00Z" } in
  check "old nik expired" (is_nik_expired old_nik);

  (* ---- sign_plan / verify_plan ---- *)
  let plan = JObj [
    ("planId", JStr "p1");
    ("startAt", JStr "2026-05-01T00:00:00Z");
    ("quantum", JInt 60)] in
  let sp = sign_plan plan nik.private_key_b64 ~kid:nik.kid () in
  check "coseSign1 protected" (String.length sp.cose_sign1.protected_hdr > 0);
  check "coseSign1 signature" (String.length sp.cose_sign1.signature > 0);
  check_eq "kid in unprotected" sp.cose_sign1.kid nik.kid;

  let cache = [(nik.kid, nik)] in
  let (ok1, r1) = verify_plan sp cache in
  check "verify ok" ok1;
  check_eq "verify reason ok" r1 "ok";

  let (ok2, r2) = verify_plan sp [] in
  check "verify fails empty cache" (not ok2);
  check_eq "reason key_not_in_cache" r2 "key_not_in_cache";

  let expired_cache = [(nik.kid, { nik with expires_at = "2000-01-01T00:00:00Z" })] in
  let (ok3, r3) = verify_plan sp expired_cache in
  check "verify fails expired key" (not ok3);
  check_eq "reason key_expired" r3 "key_expired";

  let tampered = { sp with plan = JObj [("planId", JStr "TAMPERED")] } in
  let (ok4, r4) = verify_plan tampered cache in
  check "verify fails tampered" (not ok4);
  check_eq "reason payload_mismatch" r4 "payload_mismatch";

  (* ---- SequenceTracker ---- *)
  let st = new_sequence_tracker "plan-x" in
  check_eq "plan_id stored" st.plan_id "plan-x";

  let (r1a, m1a) = add_seq st "alice" 1 in
  check "first seq accepted" r1a;
  check_eq "first seq msg" m1a "ok";

  let (r2a, m2a) = add_seq st "alice" 2 in
  check "seq 2 accepted" r2a;
  check_eq "seq 2 msg" m2a "ok";

  let (r3a, m3a) = add_seq st "alice" 2 in
  check "replay rejected" (not r3a);
  check_eq "replay msg" m3a "replay";

  let (r4a, m4a) = add_seq st "alice" 10 in
  check "gap accepted" r4a;
  check_eq "gap msg" m4a "gap";

  let (c1, _) = check_seq st "alice" 10 in
  check "check_seq replay" (not c1);
  let (c2, _) = check_seq st "alice" 11 in
  check "check_seq next" c2;
  let (c3, mc) = check_seq st "alice" 20 in
  check "check_seq gap" c3;
  check_eq "check_seq gap msg" mc "gap";

  let (rb, mb) = add_seq st "bob" 5 in
  check "bob first seq" rb;
  check_eq "bob first msg" mb "ok";

  Printf.printf "\n%d passed, %d failed\n" !passed !failed;
  if !failed > 0 then exit 1
